#!/bin/bash


# Start up local docker container of Ubuntu 21.10
docker pull ubuntu:impish
docker run -it --rm=true --platform linux/amd64 -v ~/_data:/data \
    ubuntu:impish /bin/bash

# # Then, in interactive job:

apt-get clean && \
    apt-get update -y && \
    apt-get install -y libgtk2.0-0 libgtk2.0-bin libgconf-2-4

dpkg -i /data/software/mega-cc-11.0.13-1.x86_64.deb

export THREADS=4
export TARGET=/data/phylo

export OUT_PREFIX=chir_mega
export OUT_DIR=${OUT_PREFIX}
mkdir ${OUT_DIR}
cd ${OUT_DIR}

export CONCAT_ALIGNS=mafft_aligns_concat.faa
cp ${TARGET}/${CONCAT_ALIGNS} ./

export ML_TREE=chir_ml.tree
cp ${TARGET}/${ML_TREE} ./


cat << EOF > chir_tree_mega.mao
; Please do not edit this file! If this file is modified, results are unpredictable.
; Instead of modifying this file, simply create a new MEGA Analysis Options file by using MEGA.
[ MEGAinfo ]
ver                                  = 11220624-x86_64 Linux
[ DataSettings ]
datatype                             = snProtein
MissingBaseSymbol                    = ?
IdenticalBaseSymbol                  = .
GapSymbol                            = -
Labelled Sites                       = All Sites
Labels to Include                    =
[ ProcessTypes ]
ppRelTimeML                          = true
[ AnalysisSettings ]
Analysis                             = Estimate Divergence Times (ML)
Tree to Use                          = Use tree from file
Reltime Settings                     = ====================
Clock Type                           = Local clocks
Variance Estimation Method           = Analytical method
Max Relative Rate Ratio              = 20
Statistical Method                   = Maximum Likelihood
Substitution Model                   = ====================
Substitutions Type                   = Amino acid
Model/Method                         = LG model
Rates and Patterns                   = ====================
Rates among Sites                    = Uniform Rates
No of Discrete Gamma Categories      = Not Applicable
Data Subset to Use                   = ====================
Gaps/Missing Data Treatment          = Complete deletion
Site Coverage Cutoff (%)             = Not Applicable
System Resource Usage                = ====================
Number of Threads                    = ${THREADS}
Has Time Limit                       = False
Maximum Execution Time               = -1

EOF


echo "Anopheles_stephensi=outgroup" > chir_outgroup.txt


#' From megacc documentation:
#'
#' megacc -a reltime_ML_nucleotide.mao -d mtCDNA.meg -t mtCDNA.nwk \
#'     -g mtCDNA_outgroup.txt -c mtCDNACalibrationDensities.txt \
#'     -o reltime_example_time_tree
#' The options used are:
#' -a reltime_ML_nucleotide.mao is the MEGA Analysis Options file created
#'     using MEGA GUI in prototype mode
#' -d mtCDNA.meg is a multiple sequence alignment file - a FASTA file could be
#'     used instead of a .meg file
#' -t mtCDNA.nwk is a Newick formatted tree file giving the topology of the
#'     time tree
#' -g mtCDNA_outgroup.txt is a text file that specifies which taxa in the
#'     tree comprise the outgroup - can be one or more taxa
#' -c mtCDNACalibrationDensities.txt is a text file that contains calibration
#'     density distributions for calibrating the time tree
#' -o reltime_example_time_tree is the output base file name. MEGA will
#'     produce multiple results files and their names will be prefixed with
#'     this value

megacc \
    -a chir_tree_mega.mao \
    -d ${CONCAT_ALIGNS} \
    -t ${ML_TREE} \
    -g chir_outgroup.txt \
    -o ${OUT_PREFIX}

# Takes 8805.039 sec (2 hrs 27 min)

echo rm ${ML_TREE} ${CONCAT_ALIGNS}

cp ${OUT_PREFIX}_relTimes.nwk ${TARGET}/

cd ..
tar -czf ${OUT_DIR}.tar.gz ${OUT_DIR}

mv ${OUT_DIR}.tar.gz ${TARGET}/
